lib/deep_cover/autoload_tracker.rb

```diff
@@ -0,0 +1,208 @@
+# frozen_string_literal: true
+
+require 'weakref'
+
+# TODO: if a constant is removed, AutoloadEntries should be removed
+
+module DeepCover
+  class AutoloadTracker
+    AutoloadEntry = Struct.new(:weak_const, :name, :target_path, :interceptor_path) do
+      # If the ref is dead, will return nil, otherwise the target
+      def const
+        weak_const.__getobj__
+      rescue RefError
+        nil
+      end
+    end
+
+    attr_reader :autoloads_by_basename, :interceptor_files_by_path
+    def initialize
+      @autoloads_by_basename = {}
+      @interceptor_files_by_path = {}
+    end
+
+    def autoload_path_for(const, name, path)
+      interceptor_path = setup_interceptor_for(const, name, path)
+
+      if DeepCover.custom_requirer.is_being_required?(path)
+        $LOADED_FEATURES.first
```

Comment:
Is that always the case? Wondering about cascading autoloads or something...

Reply:
The only worry I see here is if the $LOADED_FEATURES is empty... but that honestly should never happen.

What do you mean otherwise? If at any point in the current nesting of require, we are requiring that file, we must not setup an autoload. If we set it up when we go back up the nesting, to the actual file that has the definition, if it didn't yet define the constant, when it reaches it, it will try to run the autoload and that's where things go boom.

Reply:
My worry is that the right answer would be `$LOADED_FEATURES[1]`, say. Is that not possible?

Reply:
We just need ruby to believe the autoload has already been loaded. 

We can't actually remove autoload. 

So the next best thing is an entry from LOADED_FEATURES. Any of them are good enough. Unless someone's code remove entries, in which case, we would want an entry that is still there.

Reply:
Ok, cool, then a comment to that effect (e.g. "return any already loaded feature") would clarify this. Somehow I was confused and though this was the last loaded feature (not the first) so that it was assuming a particular loading order.

Reply:
Done